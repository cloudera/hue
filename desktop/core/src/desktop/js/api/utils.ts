// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import axios, {
  AxiosError,
  AxiosRequestConfig,
  AxiosResponse,
  AxiosTransformer,
  CancelToken
} from 'axios';
import qs from 'qs';

import { CancellablePromise } from './cancellablePromise';
import { hueWindow } from 'types/types';
import huePubSub from 'utils/huePubSub';
import logError from 'utils/logError';

export interface DefaultApiResponse {
  status: number;
  code?: number;
  error?: string | unknown;
  message?: string;
  responseText?: string;
  statusText?: string;
  traceback?: string;
}

export interface ApiFetchOptions<T, E = string> extends AxiosRequestConfig {
  silenceErrors?: boolean;
  ignoreSuccessErrors?: boolean;
  transformResponse?: AxiosTransformer;
  handleSuccess?: (
    response: T & DefaultApiResponse,
    resolve: (val: T) => void,
    reject: (err: unknown) => void
  ) => void;
  handleError?: (
    errorResponse: AxiosError<E>,
    resolve: (val: T) => void,
    reject: (err: unknown) => void
  ) => void;
}

const axiosInstance = axios.create({ withCredentials: true });

let baseUrl = (window as hueWindow).HUE_BASE_URL;
let bearerToken: string | undefined;

axiosInstance.interceptors.request.use(config => {
  if (baseUrl) {
    config.baseURL = baseUrl;
  }
  if (bearerToken) {
    config.headers['Authorization'] = `Bearer ${bearerToken}`;
  }
  return config;
});

axiosInstance.interceptors.response.use(response => {
  if (response.data?.access) {
    bearerToken = response.data.access;
  }
  return response;
});

export const setBaseUrl = (newBaseUrl: string): void => {
  baseUrl = newBaseUrl;
};

export const setBearerToken = (newBearerToken: string): void => {
  bearerToken = newBearerToken;
};

export const successResponseIsError = (responseData?: DefaultApiResponse): boolean => {
  return (
    typeof responseData !== 'undefined' &&
    (typeof responseData.traceback !== 'undefined' ||
      (typeof responseData.status !== 'undefined' && responseData.status !== 0) ||
      responseData.code === 503 ||
      responseData.code === 500)
  );
};

const UNKNOWN_ERROR_MESSAGE = 'Unknown error occurred';

export const extractErrorMessage = (
  errorResponse?: DefaultApiResponse | AxiosError | string
): string => {
  if (!errorResponse) {
    return UNKNOWN_ERROR_MESSAGE;
  }
  if (typeof errorResponse === 'string') {
    return errorResponse;
  }
  const defaultResponse = <DefaultApiResponse>errorResponse;
  if (defaultResponse.statusText && defaultResponse.statusText !== 'abort') {
    return defaultResponse.statusText;
  }
  if (defaultResponse.responseText) {
    try {
      const errorJs = JSON.parse(defaultResponse.responseText);
      if (errorJs.message) {
        return errorJs.message;
      }
    } catch (err) {}
    return defaultResponse.responseText;
  }
  if (errorResponse.message) {
    return errorResponse.message;
  }
  if (defaultResponse.statusText) {
    return defaultResponse.statusText;
  }
  if (defaultResponse.error && typeof defaultResponse.error === 'string') {
    return defaultResponse.error;
  }
  return UNKNOWN_ERROR_MESSAGE;
};

const notifyError = <T>(
  message: string,
  response: unknown,
  options?: Pick<ApiFetchOptions<T>, 'silenceErrors'>
): void => {
  if (!options || !options.silenceErrors) {
    logError(response);
    if (message.indexOf('AuthorizationException') === -1) {
      huePubSub.publish('hue.error', message);
    }
  }
};

const handleErrorResponse = <T>(
  err: AxiosError<DefaultApiResponse>,
  reject: (reason?: unknown) => void,
  options?: Pick<ApiFetchOptions<T>, 'silenceErrors'>
): void => {
  const errorMessage = extractErrorMessage(err.response && err.response.data);
  reject(errorMessage);
  notifyError(errorMessage, (err && err.response) || err, options);
};

const handleResponse = <T>(
  response: AxiosResponse<T & DefaultApiResponse>,
  resolve: (value?: T) => void,
  reject: (reason?: unknown) => void,
  options?: ApiFetchOptions<T>
): void => {
  if (options && options.handleSuccess) {
    options.handleSuccess(response.data, resolve, reason => {
      reject(reason);
      notifyError(String(reason), response.data, options);
    });
  } else if ((!options || !options.ignoreSuccessErrors) && successResponseIsError(response.data)) {
    const errorMessage = extractErrorMessage(response && response.data);
    reject(errorMessage);
    notifyError(errorMessage, response, options);
  } else {
    resolve(response.data);
  }
};

const getCancelToken = (): { cancelToken: CancelToken; cancel: () => void } => {
  const cancelTokenSource = axios.CancelToken.source();
  return { cancelToken: cancelTokenSource.token, cancel: cancelTokenSource.cancel };
};

export const post = <T, U = unknown>(
  url: string,
  data?: U,
  options?: ApiFetchOptions<T>
): CancellablePromise<T> =>
  new CancellablePromise((resolve, reject, onCancel) => {
    const { cancelToken, cancel } = getCancelToken();
    let completed = false;

    axiosInstance
      .post<T & DefaultApiResponse>(url, qs.stringify(data), {
        cancelToken,
        ...options
      })
      .then(response => {
        handleResponse(response, resolve, reject, options);
      })
      .catch((err: AxiosError) => {
        if (options && options.handleError) {
          options.handleError(err, resolve, reason => {
            handleErrorResponse(err, reject, options);
            notifyError(String(reason), err, options);
          });
        } else {
          handleErrorResponse(err, reject, options);
        }
      })
      .finally(() => {
        completed = true;
      });

    onCancel(() => {
      if (!completed) {
        cancel();
      }
    });
  });

export const get = <T, U = unknown>(
  url: string,
  data?: U,
  options?: ApiFetchOptions<T>
): CancellablePromise<T> =>
  new CancellablePromise((resolve, reject, onCancel) => {
    const { cancelToken, cancel } = getCancelToken();
    let completed = false;

    axiosInstance
      .get<T & DefaultApiResponse>(url, {
        cancelToken,
        params: data
      })
      .then(response => {
        handleResponse(response, resolve, reject, options);
      })
      .catch((err: AxiosError) => {
        handleErrorResponse(err, reject, options);
      })
      .finally(() => {
        completed = true;
      });

    onCancel(() => {
      if (!completed) {
        cancel();
      }
    });
  });

export const upload = async <T>(
  url: string,
  data: FormData,
  progressCallback?: (progress: number) => void
): Promise<T> => {
  const response = await axiosInstance.post<T>(url, data, {
    headers: { 'Content-Type': 'multipart/form-data' },
    onUploadProgress: progressEvent => {
      if (progressCallback) {
        progressCallback(Math.round((progressEvent.loaded * 100) / progressEvent.total));
      }
    }
  });
  return response.data;
};
