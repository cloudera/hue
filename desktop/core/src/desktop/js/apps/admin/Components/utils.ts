// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// 'License'); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an 'AS IS' BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

export const SERVER_LOGS_API_URL = '/api/v1/logs';
export const CHECK_CONFIG_EXAMPLES_API_URL = '/api/v1/check_config';
export const INSTALL_APP_EXAMPLES_API_URL = '/api/v1/install_app_examples';
export const INSTALL_AVAILABLE_EXAMPLES_API_URL = '/api/v1/available_app_examples';
export const HUE_DOCS_CONFIG_URL = 'https://docs.gethue.com/administrator/configuration/';
export const GET_USAGE_ANALYTICS_API_URL = '/api/v1/usage_analytics';
export const UPDATE_USAGE_ANALYTICS_API_URL = '/api/v1/usage_analytics/update';
