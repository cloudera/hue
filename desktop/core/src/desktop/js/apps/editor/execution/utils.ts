// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Executor from 'apps/editor/execution/executor';
import SqlExecutable from 'apps/editor/execution/sqlExecutable';
import { StatementDetails } from 'parse/types';

export interface SyncSqlExecutablesResult {
  all: SqlExecutable[];
  edited: SqlExecutable[];
  lost: SqlExecutable[];
  selected: SqlExecutable[];
  active: SqlExecutable;
}

export const syncSqlExecutables = (
  executor: Executor,
  statementDetails: StatementDetails
): SyncSqlExecutablesResult => {
  const allNewStatements = [
    ...statementDetails.precedingStatements,
    statementDetails.activeStatement,
    ...statementDetails.followingStatements
  ];

  const existingExecutables: (SqlExecutable | undefined)[] = [...executor.executables];

  const result = {
    all: <SqlExecutable[]>[],
    edited: <SqlExecutable[]>[],
    lost: <SqlExecutable[]>[],
    selected: <SqlExecutable[]>[]
  };

  let activeDatabase = executor.database();
  let currentSelectedIndex = 0;
  let activeStatementIndex = 0;
  allNewStatements.forEach((parsedStatement, index) => {
    if (/USE/i.test(parsedStatement.firstToken)) {
      const dbMatch = parsedStatement.statement.match(/use\s+([^;]+)/i);
      if (dbMatch) {
        activeDatabase = dbMatch[1];
      }
    }

    let executable = existingExecutables[index] as SqlExecutable;
    if (executable) {
      const edited =
        executable.database !== activeDatabase ||
        parsedStatement.statement !== executable.parsedStatement.statement;
      existingExecutables[index] = undefined; // undefined = not lost below
      executable.database = activeDatabase;
      executable.parsedStatement = parsedStatement;
      if (edited) {
        executable.edited = true;
        result.edited.push(executable);
      }
    } else {
      executable = new SqlExecutable({
        parsedStatement: parsedStatement,
        database: activeDatabase,
        executor: executor
      });
      executable.edited = true;
    }
    if (parsedStatement === statementDetails.activeStatement) {
      activeStatementIndex = index;
    }
    result.all.push(executable);

    if (
      currentSelectedIndex < statementDetails.selectedStatements.length &&
      parsedStatement === statementDetails.selectedStatements[currentSelectedIndex]
    ) {
      result.selected.push(executable);
      currentSelectedIndex++;
    }
  });

  result.lost = existingExecutables.filter(
    executable => typeof executable !== 'undefined'
  ) as SqlExecutable[];

  return { ...result, active: result.all[activeStatementIndex] };
};
