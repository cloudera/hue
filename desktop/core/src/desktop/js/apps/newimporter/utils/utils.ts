// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { type ColumnProps } from 'cuix/dist/components/Table';
import { ImporterFileSource, GuessFieldTypesColumn, ImporterTableData } from '../types';
import { getLastDirOrFileNameFromPath } from '../../../reactComponents/PathBrowser/PathBrowser.util';

export const convertToAntdColumns = (
  input?: GuessFieldTypesColumn[]
): ColumnProps<ImporterTableData>[] => {
  if (!input) {
    return [];
  }
  return input?.map(item => ({
    title: item.name,
    dataIndex: item.name,
    key: item.name,
    width: '100px'
  }));
};

export const convertToDataSource = (
  columns: ColumnProps<ImporterTableData>[],
  apiResponse?: string[][]
): ImporterTableData[] => {
  if (!apiResponse) {
    return [];
  }
  return apiResponse?.map((rowData, index) => {
    const row = {
      importerDataKey: `${rowData[0]}__${index}` // this ensure the key is unique
    };
    columns.forEach((column, index) => {
      if (column.key) {
        row[column.key] = rowData[index];
      }
    });
    return row;
  });
};

export const getDefaultTableName = (filePath: string, fileSource: ImporterFileSource): string => {
  // For local files, the file name is extracted from the path
  // Example: /**/**/**:fileName;**.fileExtension
  if (fileSource === ImporterFileSource.LOCAL) {
    const match = filePath.match(/:(.*?);/);
    return match?.[1] ?? '';
  }

  // For Remote, remove extension and replace '.' with '_'
  // Example: file.name.fileExtension -> file_name
  const fileName = getLastDirOrFileNameFromPath(filePath);
  if (fileName.split('.').length === 1) {
    // If there is no extension, return the file name as is
    return fileName;
  }
  return fileName.split('.').slice(0, -1).join('_');
};
