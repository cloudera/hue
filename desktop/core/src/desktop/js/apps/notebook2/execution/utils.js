// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import SqlExecutable from 'apps/notebook2/execution/sqlExecutable';

/**
 * @param executor
 * @param statementDetails
 * @param [Snippet] snippet - Optional Snippet for history
 * @return {{all: [], edited: [], lost: [], selected: []}}
 */
export const syncExecutables = (executor, statementDetails, snippet) => {
  const allNewStatements = statementDetails.precedingStatements.concat(
    statementDetails.activeStatement,
    statementDetails.followingStatements
  );

  const existingExecutables = executor.executables.concat();

  const result = {
    all: [],
    edited: [],
    lost: [],
    selected: []
  };

  let activeDatabase = executor.database();
  let currentSelectedIndex = 0;
  allNewStatements.forEach((parsedStatement, index) => {
    if (/USE/i.test(parsedStatement.firstToken)) {
      const dbMatch = parsedStatement.statement.match(/use\s+([^;]+)/i);
      if (dbMatch) {
        activeDatabase = dbMatch[1];
      }
    }

    let executable = existingExecutables[index];
    if (executable) {
      const edited =
        executable.database !== activeDatabase ||
        parsedStatement.statement !== executable.parsedStatement.statement;
      existingExecutables[index] = undefined; // undefined = not lost below
      executable.database = activeDatabase;
      executable.parsedStatement = parsedStatement;
      if (edited) {
        result.edited.push(executable);
      }
    } else {
      executable = new SqlExecutable({
        parsedStatement: parsedStatement,
        database: activeDatabase,
        executor: executor,
        snippet: snippet
      });
    }
    result.all.push(executable);

    if (
      currentSelectedIndex < statementDetails.selectedStatements.length &&
      parsedStatement === statementDetails.selectedStatements[currentSelectedIndex]
    ) {
      result.selected.push(executable);
      currentSelectedIndex++;
    }
  });

  result.lost = existingExecutables.filter(executable => typeof executable !== 'undefined');

  return result;
};
