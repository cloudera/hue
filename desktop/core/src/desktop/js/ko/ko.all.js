// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as ko from 'knockout';
import 'ko/ko.init';
import komapping from 'knockout.mapping';
import 'knockout-switch-case';
import 'knockout-sortable';
import 'knockout.validation';

import 'ext/ko.editable.custom';
import 'ext/ko.selectize.custom';

import 'ko/bindings/charts/mapchart/ko.mapChart';
import 'ko/bindings/charts/ko.barChart';
import 'ko/bindings/charts/ko.leafletMapChart';
import 'ko/bindings/charts/ko.lineChart';
import 'ko/bindings/charts/ko.partitionChart';
import 'ko/bindings/charts/ko.pieChart';
import 'ko/bindings/charts/ko.scatterChart';
import 'ko/bindings/charts/ko.timelineChart';

import 'ko/bindings/ko.vue.js';

import 'ko/bindings/ace/ko.aceEditor';
import 'ko/bindings/ace/ko.aceResizer';
import 'ko/bindings/ko.appAwareTemplateContextMenu';
import 'ko/bindings/ko.assistFileDraggable';
import 'ko/bindings/ko.assistFileDroppable';
import 'ko/bindings/ko.assistVerticalResizer';
import 'ko/bindings/ko.attachViewModelToElementData';
import 'ko/bindings/ko.augmentHtml';
import 'ko/bindings/ko.autocomplete';
import 'ko/bindings/ko.autogrowInput';
import 'ko/bindings/ko.blurHide';
import 'ko/bindings/ko.bytesize';
import 'ko/bindings/ko.chosen';
import 'ko/bindings/ko.clearable';
import 'ko/bindings/ko.clickForAceFocus';
import 'ko/bindings/ko.clickToCopy';
import 'ko/bindings/ko.clipboard';
import 'ko/bindings/ko.codemirror';
import 'ko/bindings/ko.contextMenu';
import 'ko/bindings/ko.contextSubMenu';
import 'ko/bindings/ko.datalist';
import 'ko/bindings/ko.datepicker';
import 'ko/bindings/ko.dateRangePicker';
import 'ko/bindings/ko.dblClick';
import 'ko/bindings/ko.delayedOverflow';
import 'ko/bindings/ko.dockable';
import 'ko/bindings/ko.documentChooser';
import 'ko/bindings/ko.documentContextPopover';
import 'ko/bindings/ko.draggableText';
import 'ko/bindings/ko.dropdown';
import 'ko/bindings/ko.dropzone';
import 'ko/bindings/ko.duration';
import 'ko/bindings/ko.ellipsis';
import 'ko/bindings/ko.fadeVisible';
import 'ko/bindings/ko.fetchMore';
import 'ko/bindings/ko.fileChooser';
import 'ko/bindings/ko.foreachVisible';
import 'ko/bindings/ko.fresherEditor';
import 'ko/bindings/ko.hdfsAutocomplete';
import 'ko/bindings/ko.hdfsTree';
import 'ko/bindings/ko.highlight';
import 'ko/bindings/ko.hiveChooser';
import 'ko/bindings/ko.html';
import 'ko/bindings/ko.hueach';
import 'ko/bindings/ko.hueCheckAll';
import 'ko/bindings/ko.hueCheckbox';
import 'ko/bindings/ko.hueChecked';
import 'ko/bindings/ko.hueLink';
import 'ko/bindings/ko.hueSpinner';
import 'ko/bindings/ko.impalaDagre';
import 'ko/bindings/ko.jHueRowSelector';
import 'ko/bindings/ko.logResizer';
import 'ko/bindings/ko.logScroller';
import 'ko/bindings/ko.medium';
import 'ko/bindings/ko.moment';
import 'ko/bindings/ko.momentFromNow';
import 'ko/bindings/ko.multiCheck';
import 'ko/bindings/ko.multiCheckForeachVisible';
import 'ko/bindings/ko.multiClick';
import 'ko/bindings/ko.multiLineEllipsis';
import 'ko/bindings/ko.numberFormat';
import 'ko/bindings/ko.numericTextInput';
import 'ko/bindings/ko.onClickOutside';
import 'ko/bindings/ko.oneClickSelect';
import 'ko/bindings/ko.parseArguments';
import 'ko/bindings/ko.publish';
import 'ko/bindings/ko.reactWrapper';
import 'ko/bindings/ko.readOnlyAce';
import 'ko/bindings/ko.resizable';
import 'ko/bindings/ko.select2';
import 'ko/bindings/ko.simplesize';
import 'ko/bindings/ko.slider';
import 'ko/bindings/ko.slideVisible';
import 'ko/bindings/ko.solrChooser';
import 'ko/bindings/ko.spinEdit';
import 'ko/bindings/ko.splitDraggable';
import 'ko/bindings/ko.splitFlexDraggable';
import 'ko/bindings/ko.sqlContextPopover';
import 'ko/bindings/ko.stickVertical';
import 'ko/bindings/ko.storageContextPopover';
import 'ko/bindings/ko.stretchDown';
import 'ko/bindings/ko.tagEditor';
import 'ko/bindings/ko.tagsNotAllowed';
import 'ko/bindings/ko.templateContextMenu';
import 'ko/bindings/ko.templatePopover';
import 'ko/bindings/ko.textSqueezer';
import 'ko/bindings/ko.toggle';
import 'ko/bindings/ko.toggleOverflow';
import 'ko/bindings/ko.tooltip';
import 'ko/bindings/ko.truncatedText';
import 'ko/bindings/ko.typeahead';
import 'ko/bindings/ko.verticalSlide';
import 'ko/bindings/ko.visibleOnHover';

import 'ko/components/appSwitcher/ko.appSwitcher';
import 'ko/components/assist/ko.assistDashboardPanel';
import 'ko/components/assist/ko.assistDbPanel';
import 'ko/components/assist/ko.assistDocumentsPanel';
import 'ko/components/assist/ko.assistEditorContextPanel';
import 'ko/components/assist/ko.assistFunctionsPanel';
import 'ko/components/assist/ko.assistGitPanel';
import 'ko/components/assist/ko.assistHBasePanel';
import 'ko/components/assist/ko.assistLangRefPanel';
import 'ko/components/assist/ko.assistPanel';
import 'ko/components/assist/ko.assistSchedulePanel';
import 'ko/components/assist/ko.assistStoragePanel';
import 'ko/components/assist/ko.rightAssistPanel';
import 'ko/components/contextPopover/ko.contextPopover';
import 'ko/components/contextPopover/ko.quickQueryContext';
import 'ko/components/contextPopover/ko.popoverOnHover';
import 'ko/components/doc/ko.shareDocModal';
import 'ko/components/simpleAceEditor/ko.simpleAceEditor';

import 'ko/components/ko.catalogEntriesList';
import 'ko/components/ko.contextSelector';
import 'ko/components/ko.createDirectoryModal';
import 'ko/components/ko.deleteDocModal';
import 'ko/components/ko.dropDown';
import 'ko/components/ko.editorDroppableMenu';
import 'ko/components/ko.executionAnalysis';
import 'ko/components/ko.favoriteApp';
import 'ko/components/ko.fieldSamples';
import 'ko/components/ko.globalSearch';
import 'ko/components/ko.historyPanel';
import 'ko/components/ko.importDocumentsModal';
import 'ko/components/ko.inlineAutocomplete';
import 'ko/components/ko.jobBrowserLinks';
import 'ko/components/ko.navProperties';
import 'ko/components/ko.navTags';
import 'ko/components/ko.pollingCatalogEntriesList';
import 'ko/components/ko.quickQueryAction';
import 'ko/components/ko.sentryPrivileges';
import 'ko/components/ko.sessionAuthModal';
import 'ko/components/ko.sessionPanel';
import 'ko/components/ko.shareGistModal';
import 'ko/components/ko.sqlColumnsTable';

// TODO: Move to about app when it has it's own webpack entry
import 'apps/about/components/ko.hueConfigTree';
import 'apps/about/components/ko.connectorsConfig';

import 'ko/extenders/ko.maxLength';
import 'ko/extenders/ko.numeric';
import 'ko/extenders/ko.toJson';

import 'ko/observables/ko.observableArrayDefault';
import 'ko/observables/ko.observableDefault';

window.ko = ko;
window.ko.mapping = komapping;

export default ko;
