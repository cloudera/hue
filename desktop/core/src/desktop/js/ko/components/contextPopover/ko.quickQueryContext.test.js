// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { koSetup } from 'jest/koTestUtils';
import { NAME } from './ko.quickQueryContext';
import * as hueConfig from 'utils/hueConfig';
import apiHelper from 'api/apiHelper';

import $ from 'jquery';

describe('ko.quickQueryContext.js', () => {
  const setup = koSetup();

  it('should render component', async () => {
    const connectorsSpy = jest
      .spyOn(hueConfig, 'filterEditorConnectors')
      .mockImplementation(() => [{ id: 'impala' }]);
    const computeSpy = jest
      .spyOn(apiHelper, 'fetchContextComputes')
      .mockImplementation(() => $.Deferred().resolve({}));
    const fetchContextNamespaces = jest
      .spyOn(apiHelper, 'fetchContextNamespaces')
      .mockImplementation(() => $.Deferred().resolve({}));

    const element = await setup.renderComponent(NAME, {});

    expect(connectorsSpy).toHaveBeenCalled();
    expect(computeSpy).toHaveBeenCalled();
    expect(fetchContextNamespaces).toHaveBeenCalled();

    expect(
      element.innerHTML.replace(/[0-9a-f]{8}(?:-[0-9a-f]{4}){3}-[0-9a-f]{12}/gi, 'uuid-uuid')
    ).toMatchSnapshot();

    connectorsSpy.mockRestore();
    connectorsSpy.mockClear();
  });
});
