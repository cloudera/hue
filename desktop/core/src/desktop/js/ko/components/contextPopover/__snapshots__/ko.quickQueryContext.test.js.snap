// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ko.quickQueryContext.js should render component 1`] = `
"<div data-bind=\\"descendantsComplete: descendantsComplete, component: { name: &quot;quick-query-context&quot;, params: params }\\"><div class=\\"context-popover-flex-fill\\" style=\\"overflow: auto;\\">
  <div style=\\"display: inline-block\\" data-bind=\\"
    component: {
      name: 'hue-drop-down',
      params: {
        value: connector,
        labelAttribute: 'displayName',
        entries: availableConnectors,
        linkTitle: 'Active connector'
      }
    }
  \\"><a class=\\"inactive-action hue-drop-down-active\\" href=\\"javascript:void(0)\\" data-bind=\\"toggle: dropDownVisible, css: { 'blue': dropDownVisible }\\">
    <!-- ko if: icon --><!-- /ko -->
    <!-- ko if: !noLabel && value -->
    <span class=\\"hue-drop-down-selected\\" data-bind=\\"text: value() &amp;&amp; typeof value()[labelAttribute] !== 'undefined' ? value()[labelAttribute] : value(), visible: ! dropDownVisible() || !searchable, attr: { 'title': titleTooltip }\\" title=\\"[object Object]\\">[object Object]</span>
    <!-- /ko -->
    <i class=\\"fa fa-caret-down\\"></i>
  </a><input class=\\"hue-drop-down-input\\" type=\\"text\\" data-bind=\\"textInput: filter, attr: { 'placeHolder': inputPlaceHolder }, visible: dropDownVisible, style: { color: filterEdited() ? '#000' : '#AAA', 'min-height': '22px', 'margin-left': '10px' }\\" style=\\"display: none; color: rgb(170, 170, 170); min-height: 22px; margin-left: 10px;\\"><i class=\\"fa fa-caret-down\\"></i><div class=\\"hue-drop-down-container\\" data-bind=\\"css: { 'open' : dropDownVisible, 'hue-drop-down-fixed': fixedPosition, 'hue-drop-down-container-searchable': searchable }, dropDownKeyUp: { onEsc: onEsc, onEnter: onEnter, dropDownVisible: dropDownVisible }\\">
    <div style=\\"overflow-y: auto;\\" class=\\"dropdown-menu\\" data-bind=\\"visible: filteredEntries().length > 0\\">
      <!-- ko if: foreachVisible --><!-- /ko -->
      <!-- ko ifnot: foreachVisible -->
      <ul class=\\"hue-inner-drop-down\\" data-bind=\\"foreach: filteredEntries\\">
        <!-- ko if: typeof $data.divider !== 'undefined' && $data.divider --><!-- /ko -->
        <!-- ko if: typeof $data.divider === 'undefined' || !$data.divider -->
        <li><a href=\\"javascript:void(0)\\" data-bind=\\"text: $data &amp;&amp; typeof $data[$parent.labelAttribute] !== 'undefined' ? $data[$parent.labelAttribute] : $data, click: function () { let previous = $parent.value(); $parent.value($data); $parent.onSelect($data, previous); }\\">[object Object]</a></li>
        <!-- /ko -->
      </ul>
      <!-- /ko -->
    </div>
  </div></div>
  <!-- ko if: connector() -->
    <div class=\\"margin-left-10\\" style=\\"display: inline-block\\" data-bind=\\"
      component: {
        name: 'hue-context-selector',
        params: {
          connector: connector,
          compute: compute,
          namespace: namespace,
          availableDatabases: availableDatabases,
          database: database,
          hideLabels: true
        }
      }
    \\"><i class=\\"fa fa-spinner fa-spin muted\\"></i><div class=\\"inline-block\\" style=\\"display: none;\\" data-bind=\\"visible: !loadingContext()\\">
    <!-- ko if: window.HAS_MULTI_CLUSTER --><!-- /ko -->

    <!-- ko if: availableDatabases().length > 0 && !hideDatabases--><!-- /ko -->
    <!-- ko if: availableDatabases().length === 0  && !hideDatabases -->
    <span class=\\"editor-header-title\\"><i class=\\"fa fa-warning\\"></i> No databases found</span>
    <!-- /ko -->
  </div></div>
  <!-- /ko -->
  <!-- ko ifnot: loadingContext --><!-- /ko -->
</div></div>"
`;
